src/effects/backends/builtin/pitchshifteffect.cpp

```diff
@@ -89,28 +137,82 @@ void PitchShiftEffect::processChannel(
         }
     }();
 
-    pState->m_pRubberBand->setPitchScale(pitch);
+    if (pitch != m_prevPitch) {
+        pState->m_pRubberBand->setPitchScale(pitch);
+        m_prevPitch = pitch;
+    }
 
     SampleUtil::deinterleaveBuffer(
             pState->m_retrieveBuffer[0],
             pState->m_retrieveBuffer[1],
             pInput,
-            engineParameters.framesPerBuffer());
-    pState->m_pRubberBand->process(
-            pState->m_retrieveBuffer,
-            engineParameters.framesPerBuffer(),
-            false);
-
-    SINT framesAvailable = pState->m_pRubberBand->available();
-    SINT framesToRead = math_min(
-            framesAvailable,
-            engineParameters.framesPerBuffer());
-    SINT receivedFrames = pState->m_pRubberBand->retrieve(
-            pState->m_retrieveBuffer,
-            framesToRead);
-
-    SampleUtil::interleaveBuffer(pOutput,
-            pState->m_retrieveBuffer[0],
-            pState->m_retrieveBuffer[1],
-            receivedFrames);
+            framesPerBuffer);
+
+    SINT offsetFrames = 0;
+
+    while (offsetFrames < framesPerBuffer) {
+        SINT blockFrames = kBlockFrames;
+
+        if (offsetFrames + blockFrames > framesPerBuffer) {
+            blockFrames = framesPerBuffer - offsetFrames;
+        }
+
+        SINT processedFrames = 0;
+
+        while (processedFrames < blockFrames) {
+            // RubberBand works with multichannel Sample FRAMES.
+            SINT requiredFrames = pState->m_pRubberBand->getSamplesRequired();
+            SINT toProcessFrames = math_min(blockFrames - processedFrames, requiredFrames);
+
+            for (int c = 0; c < channelCount; ++c) {
+                pState->m_inputSamples[c] =
+                        pState->m_retrieveBuffer[c] +
+                        offsetFrames + processedFrames;
+            }
+
+            pState->m_pRubberBand->process(
+                    pState->m_inputSamples,
+                    toProcessFrames,
+                    false);
+
+            processedFrames += toProcessFrames;
+
+            SINT availableFrames = pState->m_pRubberBand->available();
+            SINT writableFrames = pState->m_outputBuffer->getWriteSpace() / channelCount;
+
+            if (availableFrames > writableFrames) {
+                qDebug() << "PitchShiftEffect::processChannel: buffer is not large enough";
```

Comment:
We must not qDebug() in the engine thread, because locking buffer allocation is involved.